addons/mail/models/mail_channel.py

```diff
@@ -322,40 +322,48 @@ def add_members(self, partner_ids=None, guest_ids=None, invite_to_rtc_call=False
+            new_members = self.env['mail.channel.partner'].sudo().create(members_to_create)
+            for channel_partner in new_members.filtered(lambda channel_partner: channel_partner.partner_id):
+                user = channel_partner.partner_id.user_ids[0] if channel_partner.partner_id.user_ids else self.env['res.users']
+                # notify invited members through the bus
+                if user:
+                    self.env['bus.bus'].sendone((self._cr.dbname, 'res.partner', channel_partner.partner_id.id), {
+                        'type': 'mail.channel_joined',
+                        'payload': {
+                            'channel': channel_partner.channel_id.with_user(user).with_context(allowed_company_ids=user.company_ids.ids).sudo().channel_info()[0],
+                            'invited_by_user_id': self.env.user.id,
+                        },
+                    })
+                # notify existing members with a new message in the channel
+                if channel_partner.partner_id == self.env.user.partner_id:
+                    notification = _('<div class="o_mail_notification">joined the channel</div>')
+                else:
+                    notification = _(
+                        '<div class="o_mail_notification">invited <a href="#" data-oe-model="res.partner" data-oe-id="%(new_partner_id)d">%(new_partner_name)s</a> to the channel</div>',
+                        new_partner_id=channel_partner.partner_id.id,
+                        new_partner_name=channel_partner.partner_id.name,
+                    )
+                channel_partner.channel_id.message_post(body=notification, message_type="notification", subtype_xmlid="mail.mt_comment", notify_by_email=False)
+                channel_data['members'][0][1].append({
+                    'id': channel_partner.partner_id.id,
+                    'im_status': channel_partner.partner_id.im_status,
+                    'name': channel_partner.partner_id.name,
+                })
+            for channel_partner in new_members.filtered(lambda channel_partner: channel_partner.guest_id):
+                channel_partner.channel_id.message_post(body=_('<div class="o_mail_notification">joined the channel</div>'), message_type="notification", subtype_xmlid="mail.mt_comment", notify_by_email=False)
+                channel_data['guestMembers'][0][1].append({
+                    'id': channel_partner.guest_id.id,
+                    'name': channel_partner.guest_id.name,
+                })
+            self.env['bus.bus'].sendone((self._cr.dbname, 'mail.channel', channel.id), {
                 'type': 'mail.channel_update',
-                'payload': {
-                    'id': channel_partner.channel_id.id,
-                    'memberCount': channel_partner.channel_id.member_count,
-                    'members': [('insert', new_partner_data)],
-                },
+                'payload': channel_data,
```

Comment:
so if possible here avoid diff and keep creating the payload in-place, just have 2 lists that you maintain